### PR TITLE
Refactor `FreeFermion2D`

### LINKED ISSUE #33: Make `a=1` the default for plain derivatives
Currently, the default for `L` is set to `1` because that's convenient to do. But traditionally one would choose `a` to be one not `L`, so we should adjust for that.

### LINKED ISSUE #35: Refactor 2D example into numpy broadcasting style
Most of the functions in `FreeFermion2D` can be significantly reduced (if not made oneliners) by leaning more on numpy's broadcasting magic. We should do so in a future revision.

### LINKED ISSUE #37: Better constructor argument structure for 2D
There is a suggestion [here](https://github.com/Mayank447/Pseudo-Spectral-Discretization/pull/9#discussion_r1696979744).

### PR DESCRIPTION
**Pull request by chillenzer**
This is a major refactoring of the `FreeFermion2D` class using a more array-focused numpy-style approach that should conveniently generalise to other dimensions.

- [x] Depends on #44 to be merged.

Closes #33.
Closes #35.
Closes #37.

Edit:

This PR combines a number of conceptually related (but admittedly distinct) changes that could be split up into separate PRs if that's easier to review for you:
- Adjustment of the interface of `FreeFermion2D` joining related arguments into collections and mandating explicit kwargs.
- Refactoring of the multiple initialisation steps into smaller methods.
- Implicit computation of spectral structure in spinor space.
- Joining `p` and `eta` into collections.
- Leveraging numpy's broadcasting instead of explicit handling of the degrees of freedom.

The above changes entail the following minor qualitative improvements of the code:
- The code gets terser. That means less code to maintain (e.g. handling of DoFs is delegated to numpy) but, of course, denser functionality and more to understand per line. That's a trade-off that by itself would potentially not justify the change.
- It fixes a few mistakes around the broadcasting you implemented. I've added the `test_broadcasting.py` to check. (The tests runs with the `main` branch as well if you replace `total_num_of_dof` with `total_num_lattice_points`).
- It is about 20% faster (on my local workstation). I've added a benchmark to verify that, so feel free to play around.

The main reason for the proposed changes is, however, that I have removed all references to the fact that it's 2D in the internals. The plan would be that the current implementation should be well-suited to be generalised to the N-dimensional problem and even to subsume the `Derivative1D` case. That would be a great feature jump and maintenance relieve. As it's an independent concern and not yet tested, I didn't bring that into the current PR.

This PR is structured as follows:
- The main development has obviously happened in `pseudospectral/spectra/fermion2D.py`. You might want to start your review here.
- Due to interface changes some tests (mostly `conftest.py`) needed adjustment.
- Added a few further tests and a benchmark.
- The changes from #44 are currently cluttering the diffs. Please merge that one first and I'll rebase, so we have it much cleaner.

### CODE REVIEW

**Review by Mayank447 — COMMENTED** [review]
Though the refactoring may be shorter. But I believe this change makes debugging and understanding what's going on incomprehensible

File: pseudospectral/spectra/fermion2D.py
Comment (Mayank447) [inline pseudospectral/spectra/fermion2D.py:1]: Though the refactoring of fft2 may be shorter. But I believe this change makes debugging and understanding what's going on incomprehensible

**Comment by chillenzer** [comment]
You are right, I'm very sorry. I was in a bit of a hurry when I submitted the PR and provided insufficient context and information for you to review. I have updated the description to how a proper PR should look in my opinion.

You are also correct about there being a trade-off between generality and readability (although being used to this array-programming-like style I find it much more readable than the previous code). For me personally, this trade-off is very well justified and I have properly laid out my arguments above. But I'm happy to extract variables and the like to make the functional code shorter and easier on the eye. Also, I think your decision to stay with the 250 character limit doesn't exactly aid readability and you might want to reconsider it after this PR is through (please don't in between due to all the merge conflicts it would create).

Happy to discuss it in the office tomorrow in more detail.

File: pseudospectral/spectra/fermion2D.py [hunk 1]
```diff
@@ -127,27 +127,16 @@ def eigenfunction(self, index):
...
             raise ValueError(f"Index {index} out of bounds.")
 
-        sign = 1 - 2 * (index % 2)
-        p_t = self.p_t[index // 2]  # Since two eigenvalues exist due to spinor structure
-        p_x = self.p_x[index // 2]
-        p_t_mu = self.p_t_mu[index // 2]
-        norm_p = self.norm_p[index // 2]
+        # Since two eigenvalues exist due to spinor structure
+        spacetime_index = index // self.dof_spinor
+        eta = self.eta.reshape(-1, self.dof_spinor, self.dof_spinor)[spacetime_index, :, index % self.dof_spinor]
```
Comment (Mayank447) [inline pseudospectral/spectra/fermion2D.py]: reshape is not necessary

File: pseudospectral/spectra/fermion2D.py [hunk 2]
```diff
@@ -118,16 +118,7 @@ def _return_eigenfunction(self, t, x, eta, p):
...
-
-        num_eigenfunction = p.shape[-1]
-        ret = np.zeros((num_eigenfunction, self.total_num_lattice_points), dtype=np.complex128)
-
-        # Kronecker product between each spinor array and corresponding exponential part
-        for i in range(num_eigenfunction):
-            ret[i] = np.kron(exp_component[i], eta[i])
-
-        return ret
+        return np.einsum("j...,jk->j...k", np.exp(np.einsum("ij,i...->j...", p, np.asarray([t, x]))).reshape(*p.shape[1:], *x.shape) / np.sqrt(np.prod(self.L)), eta)
```
Comment (Mayank447) [inline pseudospectral/spectra/fermion2D.py]:
```suggestion 
        return np.einsum("j...,jk->j...k", np.exp(np.einsum("ij,i...->j...", p, np.asarray([t, x]))).reshape(*p.shape[1:], *x.shape)), eta) / np.sqrt(np.prod(self.L)
```

File: pseudospectral/spectra/fermion2D.py [hunk 3]
```diff
@@ -110,15 +110,12 @@ def eigenfunction(self, index):
...
-        return lambda t, x: self._return_eigenfunction(t, x, eta, self.p.reshape(self.dimension, -1)[:, spacetime_index])
-
-    def _return_eigenfunction(self, t, x, eta, p):
-        """
-        Return function when eigenfucntion method is called.
-        """
-        return np.einsum("j...,jk->j...k", np.exp(np.einsum("ij,i...->j...", p, np.asarray([t, x]))).reshape(*p.shape[1:], *x.shape) / np.sqrt(np.prod(self.L)), eta)
+        return lambda t, x: np.einsum(
+            "j...,jk->j...k",
+            np.exp(np.einsum("ij,i...->j...", self.p.reshape(self.dimension, -1)[:, spacetime_index], np.asarray([t, x]))).reshape(*index.shape, *x.shape) / np.sqrt(np.prod(self.L)),
```
Comment (Mayank447) [inline pseudospectral/spectra/fermion2D.py]: Add few temporary variables with appropriate names

File: benchmark/fermion2D.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+for linear_num_lattice_points in [16, 32, 64, 128, 256]:
+    spectrum = create_spectrum(linear_num_lattice_points)
+    shape = (num_input_vectors, total_num_of_dof(spectrum))
+    timings[linear_num_lattice_points] = np.mean(
+        timeit.repeat(
+            'spectrum.transform(input_vectors, "real","spectral")',
+            setup="input_vectors=np.random.rand(*shape)",
+            number=1,
+            repeat=2,
+            globals=globals(),
```
Comment (Mayank447) [inline benchmark/fermion2D.py:37]: Remove globals(), be explicit

**Comment by Mayank447** [comment]
Remove the redundant docstrings, update docstrings explaining your current implementation,

**Comment by Mayank447** [comment]
Change the character limit to 100 characters (or something more reasonable)

File: pseudospectral/spectra/fermion2D.py (hunk 1 above)
Comment (chillenzer) [reply pseudospectral/spectra/fermion2D.py]: Done.

File: pseudospectral/spectra/fermion2D.py (hunk 2 above)
Comment (chillenzer) [reply pseudospectral/spectra/fermion2D.py]: Was already the case in latest version.

File: pseudospectral/spectra/fermion2D.py (hunk 3 above)
Comment (chillenzer) [reply pseudospectral/spectra/fermion2D.py]: Done.

File: benchmark/fermion2D.py (hunk 4 above)
Comment (chillenzer) [reply benchmark/fermion2D.py:37]: Done.

**Comment by chillenzer** [comment]
All done as far as I can see.

**Comment by Mayank447** [comment]
Everything looks good!

**Pull request merged**